components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Activity.kt

```diff
@@ -45,15 +36,27 @@ fun Activity.exitImmersiveModeIfNeeded() {
         // We left immersive mode already.
         return
     }
+    window.decorView.viewTreeObserver?.removeOnWindowFocusChangeListener(this.onWindowFocusChangeListener)
 
     window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-        window.insetsController?.show(WindowInsets.Type.systemBars())
-    } else {
-        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_VISIBLE
+    window.getWindowInsetsController().apply {
+        show(WindowInsetsCompat.Type.systemBars())
     }
 }
 
+/**
+ * OnWindowFocusChangeListener used to ensure immersive mode is not broken by other views interactions.
+ */
+@VisibleForTesting
+internal val Activity.onWindowFocusChangeListener: ViewTreeObserver.OnWindowFocusChangeListener
+    get() = ViewTreeObserver.OnWindowFocusChangeListener { hasFocus ->
+        if (hasFocus) {
+            window.getWindowInsetsController().apply {
+                hide(WindowInsetsCompat.Type.systemBars())
+            }
+        }
+    }
```

Comment:
Doesn't this return a new one every time? So do we really remove the listeners we add? 🤔

Reply:
Yeah, not sure why I thought of it as a member. 😴